h3i/src/client/connection_summary.rs

```diff
@@ -106,21 +115,105 @@ impl StreamMap {
     /// use quiche::h3::Header;
     /// use std::iter::FromIterator;
     ///
+    /// let mut stream_map = StreamMap::default();
+    ///
     /// let h = Header::new(b"hello", b"world");
     /// let headers = H3iFrame::Headers(EnrichedHeaders::from(vec![h]));
     ///
     /// let stream_map: StreamMap = [(0, vec![headers.clone()])].into();
-    /// assert_eq!(stream_map.all_frames(), vec![headers]);
+    /// assert_eq!(stream_map.stream(0), vec![headers]);
     /// ```
-    pub fn all_frames(&self) -> Vec<H3iFrame> {
-        self.0
+    pub fn stream(&self, stream_id: u64) -> Vec<H3iFrame> {
+        self.map.get(&stream_id).cloned().unwrap_or_default()
+    }
+
+    /// Collect all frames that match the predicate on a given stream.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use h3i::client::connection_summary::StreamMap;
+    /// use h3i::frame::EnrichedHeaders;
+    /// use h3i::frame::H3iFrame;
+    /// use quiche::h3::frame::Frame;
+    /// use quiche::h3::Header;
+    /// use std::iter::FromIterator;
+    ///
+    /// let h = Header::new(b"hello", b"world");
+    /// let correct_headers = H3iFrame::Headers(EnrichedHeaders::from(vec![h]));
+    /// let data = H3iFrame::QuicheH3(Frame::Data {
+    ///     payload: b"nyj".to_vec(),
+    /// });
+    /// let stream_map: StreamMap = [
+    ///     (0, vec![correct_headers.clone(), data]),
+    ///     (4, vec![correct_headers.clone()]),
+    /// ]
+    /// .into();
+    ///
+    /// assert_eq!(
+    ///     stream_map.filter_stream(0, |frame| {
+    ///         match frame {
+    ///             H3iFrame::Headers(_) => true,
+    ///             _ => false,
+    ///         }
+    ///     }),
+    ///     vec![correct_headers]
+    /// );
+    /// ```
+    pub fn filter_stream<F>(&self, stream_id: u64, f: F) -> Vec<H3iFrame>
+    where
+        F: Fn(&H3iFrame) -> bool,
+    {
+        self.stream(stream_id)
+            .into_iter()
+            .filter(|frame| f(frame))
+            .collect()
+    }
+
+    /// Collect all frames that match the predicate. The ordering is
+    /// non-deterministic.
```

Comment:
The non-deterministic bit seems annoying. If we swap the underlying collection for a BTreeMap, they would at least be consistently in increasing stream ID order.   WDYT?

Reply:
That works. It's also worth noting (and my fault for not being clear in the docs) that the per-stream frame ordering is deterministic - the order in which we iterate over the streams themselves is non-deterministic